File: src/java/magma/app/compile/lang/ForwardsLocator.java
package magma.app.compile.lang;

import magma.app.compile.rule.Locator;

import java.util.stream.Stream;
import java.util.ArrayList;
import java.util.List;

public record ForwardsLocator(String delimiter) implements Locator {
    @Override
    public String slice() {
        return delimiter;
    }

    @Override
    public Stream<Integer> locate(String input) {
        List<Integer> indices = new ArrayList<>();
        int index = input.indexOf(delimiter);

        while (index != -1) {
            indices.add(index);
            index = input.indexOf(delimiter, index + 1);
        }

        return indices.stream();
    }
}

File: src/c/magma/app/locate/ForwardsLocator.c
import magma.api.stream.Stream;import magma.api.stream.Streams;import magma.app.rule.locate.Locator;import java.util.ArrayList;import java.util.List;struct ForwardsLocator{
	String infix;
	public ForwardsLocator(any* _ref_, String infix){
		this.infix =infix;
	}
	String unwrap(){
		return this.infix;
	}
	int length(){
		return this.infix.length();
	}
	Stream<Integer> locate(any* _ref_, String input){
		return Streams.from(searchForIndices(input));
	}
	List<Integer> searchForIndices(any* _ref_, String input){
		List<Integer> indices=new ArrayList<>();
		int index=input.indexOf(this.infix);
		while(index>=0){
			indices.add(index);
			index=input.indexOf(this.infix, index+1);
		}
		return indices;
	}
	Locator N/A(){
		return N/A.new();
	}
}